File: day25/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item == '-d' ? 1 : item == '-d2' ? 2 : item == '-d3' ? 3 : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) =>
    item == '-i' ? 'input.txt' : item == '-e' ? 'example.txt' : item.startsWith('-e') ? `example${item.slice(-1)}.txt` : acc,
  'example.txt'
);
console.log({ FILENAME, DEBUG });

const text = await Deno.readTextFile(FILENAME);
if (DEBUG > 1) console.debug({ text });

const sections = text.split('\n\n');
const locks = [];
const keys = [];
const rowCount = 7;
const colCount = 5;

for (const section of sections) {
  // row->col
  const rows = section.split('\n').map((line) => line.split(''));
  // // col->row
  // const cols = rows[0].map((_, i) => rows.map((row) => row[i]));
  const heights = rows[0].map((_, colIx) =>
    rows.reduce((acc, _, rowIx) => (rows[rowIx][colIx] == (rows[0][0] == '#' ? '#' : '.') ? Math.max(acc, rowIx) : acc), 0)
  );
  if (rows[0][0] == '#') {
    locks.push(heights);
  } else {
    keys.push(heights.map((i) => rowCount - i - 2));
  }
}
if (DEBUG > 0) {
  console.debug({ locks });
  console.debug({ keys });
}

let fitCount = 0;
for (const lock of locks) {
  for (const key of keys) {
    const fit = lock.reduce((acc, _, col) => acc && key[col] + lock[col] + 1 < rowCount, true);
    if (DEBUG > 0) {
      console.debug({ lock, key, fit });
    }
    if (fit) fitCount++;
  }
}
console.log('part 1', fitCount);

File: day25/solution.py
#!/usr/bin/env python3
import sys
from functools import cache

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'example.txt'

for arg in sys.argv[1:]:
  if arg == '-i': FILENAME = 'input.txt'
  elif arg == '-e': FILENAME = 'example.txt'
  elif arg.startswith('-e'): FILENAME = f'example{arg[-1]}.txt'
  elif arg == '-d': DEBUG = 1
  elif arg == '-d2': DEBUG = 2
  elif arg == '-d3': DEBUG = 3
  else: raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  text = f.read()
locks=[]
keys=[]
cols=5
rows=7
for item in [x.splitlines() for x in text.split('\n\n')]:
  print(f'{item=}')
  if item[0][0]=='#':
    locks.append(tuple(
      max(*[r for r in range(rows) if item[r][c]=='#'],0)
      for c in range(cols)
    ))
  else:
    # lock
    keys.append(tuple(
      rows-min(*[r for r in range(rows) if item[r][c]=='#'],rows)-1
      for c in range(cols)
    ))

if DEBUG>1:
  print(f'{locks=}')
  print(f'{keys=}')
fit_count=0
for lock in locks:
  for key in keys:
    fit=True
    for c in range(cols):
      if lock[c]+key[c]>=rows-1:
        fit=False
        break
    if fit:
      fit_count+=1
      if DEBUG>0: print(f'{lock=} {key=} fit')
    else:
      if DEBUG>0: print(f'{lock=} {key=} no fit')
print(f'part 1: {fit_count=}')
